### DISCUSSION TITLE
Problem regarding Analyzability of Failing Test Cases

### DISCUSSION THREAD

**Discussion by z002Holpp** [nunit/nunit#4326] [Q&A]
Hi NUnit Community,

I would like to bring up a request for discussion:

We are coming from an R&D organization using a quite old test framework, which is based on a highly customized fork of NUnit2. We have a huge test base, speaking of multiple 100k tests that are distributed over all test levels from Unit Test to UI-based System Test. Our goal is to migrate now from this legacy test framework to vanilla NUnit3.

One of the main drivers in our test environment is analyzability, meaning that once a test case has failed, we want to ensure that it is easily possible to analyze and identify the root cause of the failure in order to find appropriate counter measures. With our old test framework, we use custom event handlers for this purpose. We have event handlers to create screenshots, environment dumps, collect product logs and much more. In order to snapshot the correct system state, it is absolutely crucial that the eventhandler is triggered in exact the moment that the test has failed. In our current test framework we have ensured this by a custom implementation of the whole event handling.

We prototyped now a lot with NUnit3, especially with custom event handlers. One of our concerns is that event handling in NUnit3 seems asynchronous by design. That means that while a custom event handler is doing its task, already the next test cases could be started and running and the system state which we want to capture is altered. And thus, our main requirement (event handler is triggered in exact the moment that the test failed) is not met.
We also evaluated other possibilities like ActionAttributes. However we see here the constraint that we cannot attribute our complete test base. Attributes – even on assembly level – would hurt us tremendously during the migration.

Would you have any proposals for us here? We highly appreciate your thoughts and inputs!

Thanks and Best regards!

Hint: Initially we started this topic in the discussion area ([link](https://github.com/nunit/nunit/discussions/4154)) but didn't get any feedback. This topic becomes more and more critical for us, so we try to get some inputs here.

**Comment by z002Holpp** [comment]
@CharliePoole @rprouse 
Do you maybe have an opinion about that topic? It would be great to get into a technical discussion with someone. :)

**Comment by rprouse** [comment]
I am going to start by labeling this issue as a question. We can transform to another state after we investigate.

Can you give specific code examples of how you have attempted to plug into the test events in NUnit 3 so that we can dig deeper into the code and come up with solutions or workarounds? There are two large feature changes that might be causing the behaviour you are describing, first is parallelization and second is multiple asserts. With code examples, we'll be able to narrow down the code we are looking at.

From your description, you are looking to be notified immediately on test failure presumably from within our Asserts.

1. Are there additional events that you are interested in through the test discovery, setup, running and finishing states?
2. For test failures, if your developers start using the multiple assert feature, when would you expect the event to fire? With each failing assert or at the end of the multiple assert block?

**Comment by CharliePoole** [comment]
@z002Holpp In reading this, I assumed that "custom event handlers" referred to engine extensions rather than catching events in the framework. Since each type is implemented differently and in different software packages, it would be important to know. That said, engine event handlers are entirely asynchronous and have to remain that way so this discussion is in the right place.

@rprouse As you know, we removed any ability to handle events synchronously in the framework with NUnit 3. The intention was that events could observe and record what happened in the execution of the test but could not influence it. In fact, we put a lot of work into making impossible for any observer of events to change the result of the test.

On the other hand, there are other facilities for extensibility, such as custom commands, which do operate synchronously. I think your request for detail is a good start, but I hope you'll leave events asynchronous.

**Comment by rprouse** [comment]
> On the other hand, there are other facilities for extensibility, such as custom commands, which do operate synchronously. I think your request for detail is a good start, but I hope you'll leave events asynchronous.

@CharliePoole my intention is only to find out exactly what they are trying to do to provide them with a solution, not to change the framework or engine.

**Comment by rprouse** [comment]
@z002Holpp, you also stated,

> We also evaluated other possibilities like ActionAttributes. However we see here the constraint that we cannot attribute our complete test base. Attributes – even on assembly level – would hurt us tremendously during the migration.

I assume that it would "hurt" your migration because you have so many test assemblies? Couldn't you write a utility to inject your custom attribute into each test project in each solution and automate the processes of migration? With a test-base this large, you are going to likely need to do similar tasks with other custom attributes. For example, you will likely need to write a custom `ExpectedExceptionAttribute` and add a reference to it.

**Comment by z002Holpp** [comment]
[ActualOutput.txt](https://github.com/nunit/nunit/files/9766475/ActualOutput.txt)
[TestEventHandlerDemo_forIssue.zip](https://github.com/nunit/nunit/files/9775543/TestEventHandlerDemo_forIssue.zip)

Hi,

first of all, thanks a lot for your replies! 

We provided a small sample solution with a custom event handler that represents the tests in our organization. The attached log shows the output of the asynchronous event handling. In case of an exception in the test, we start a savelog. This savelog is very important for us because it collects data from the loaded patient. 

From the attached logs you can see that the savelog collection is started too late:

- First, because its started after the TearDown, which itself runs for some seconds.

- Secondly, because during collecting the savelog, already the subsequent testcase has started, which alters the system state and invalidates the savelog.

What we would need is a possibility to overcome the above mentioned points such that the actions like savelog creation is started **before** the TearDown and such that the second test case would **not be started before** the savelog collection has completed.

Is there any way to achieve this?

Thanks and best regards!

**Comment by siprbaum** [comment]
Inlining the relevant files from the attached zip file to make it easier to discuss.
It contains an EventHandler like this:

```csharp
namespace MyEventHandler
{
    [Extension]
    public class MyTestEventListener : ITestEventListener
    {
        public void OnTestEvent(string report)
        {
            //a test has finished
            if (report.StartsWith("<test-case"))
            {
                //and was failing
                if (report.Contains("result=\"Failed\""))
                {
                    Logger.Log("TC Failed for :" + report);
                    ProductHelper.CreateScreenshot();
                    ProductHelper.FetchSavelogsOfRunningSystem();
                }
            }
        }
    }
}
```
and some demo test code like this

```csharp
namespace DemoTests
{
    public class Tests
    {
        private int _sleepTime = 200;

        [OneTimeSetUp]
        public void StartApplication()
        {
            Logger.Clear();
            Thread.Sleep(_sleepTime);
            Logger.Log("Application started.");
        }

        [SetUp]
        public void LoadPatient() 
        {
            Thread.Sleep(_sleepTime);
            Logger.Log("Patient loaded.");
        }

        [TearDown]
        public void ClosePatient() 
        {
            Thread.Sleep(_sleepTime);
            Logger.Log("Patient closed.");
        }

        [OneTimeTearDown]
        public void ShutdownApplication()
        {
            Thread.Sleep(_sleepTime);
            Logger.Log("Application shut down.");
        }

        [Test]
        public void TestDataLoading1()
        {
            Logger.Log("Loading Dataset1");
            Thread.Sleep(TimeSpan.FromSeconds(1));
            
            throw new Exception("some error occured.");   // <---- HERE THE TEST FAILS
            
            Thread.Sleep(TimeSpan.FromSeconds(1));
        }

        [Test]
        public void TestDataLoading2()
        {
            Logger.Log("Loading Dataset2");
            Thread.Sleep(TimeSpan.FromSeconds(2));
        }
    }
}
```

From your description, it is not 100% clear where you expect a hooking mechanism to execute custom code.
Let me try to describe:
```
StartApplication    (OneTimeSetup)
LoadPatient         (Setup)
TestDataLoading1    (Test)   < --- This one fails, as it throws an Exception
  > (1) Should the hook to your custom code (taking screenshots, savelogs etc.) be here?
ClosePatient        (TearDown)
   > (2) Or should it be after the TearDown ?
ShutdownApplication (OneTimeTearDown)
```
@z002Holpp So do you expect it to be right after the test method (1) or after the TearDown (2) ?

**Comment by CharliePoole** [comment]
In either case (1) or (2) it simply isn't possible to do this in an engine extension because they are asynchronous by design and have to remain so because the engine is generally running in a separate process (separate machine in the  future) and is handling multiple agent processes. Introducing a wait state in the tests, while arbitrary engine extensions are processing would create a point of failure outside the control of the test writer.

Normally, logging of individual tests is done in the teardown itself, usually before resources are released. This can even be generalized in a base class used for all those test fixtures which need to take screenshots.

**Comment by z002Holpp** [comment]
@siprbaum 

thanks for the inlining and highlighting the unclear spot!

Our goal is to execute the custom handlers as close as possible to the exception!

In our test environment, the teardown would often do quite expensive operations, like closing patient data that was previously loaded. Those operations can take time, sometimes multiple seconds or even up to a minute. Having the custom handler executed after the teardown means that we lose this time, the system state will change and the failure screenshot and savelog will not precisely capture the failure state. In fact, every additional time between the actual exception that leads the test case to fail and the trigger of the custom evenhandler would invalidate our captured data more and would make failure analysis more difficult.

@CharliePoole 
I got your point. However if we add the following base class for collecting the error information:

```csharp
 public class TestBase
    {
        [TearDown]
        public void CollectErrorInformation()
        {
            Logger.Log("Base Class Error Info Collection.");
        }
    }
```
then the TearDown of the base class is executed *after* the TearDown of the derived class, as you see also in the log:

 <pre><code>
10:23:01.617:	 Application started.
10:23:01.835:	 Patient loaded.
10:23:01.838:	 Loading Dataset1
<b>10:23:03.063:	 <i>Tear Down</i>: Patient closed.                            <-- Wrong order for our use case    
10:23:03.065:	 <i>Base Class Tear Down</i>: Error Info Collection.          <-- Wrong order for our use case</b>
10:23:03.281:	 Patient loaded.
10:23:03.283:	 Loading Dataset2
10:23:05.504:	 Patient closed.
10:23:05.505:	 Base Class Error Info Collection.
10:23:05.708:	 Application shut down.
</code>
</pre>

So the ordering is not working for our use case. When the Error info collection starts, we would need the data from the loaded patient. But since the TearDown of the derived class is executed before the TearDown of the base class, the needed patient data is already unloaded when the savelog collection starts.

**Comment by CharliePoole** [comment]
That's correct, but only if your code unloads the data in the derived TearDown. The order of the two teardowns is fixed, but it is you who determines what happens in each of them.

In fact, the same thing is true in a single teardown, where you order the actions in that way you write the code. So, to save the screenshot before closing the patient, your code would be in that order...

```
SaveScreenShot();
ClosePatient();
```

If all your teardowns end with that sequence, then you can save keystrokes by use of a base-class teardown, which contains them.

**Comment by z002Holpp** [comment]
SaveScreenshot() is only an example for methods that collect data which we need to analyze failed tests. Those methods can be seen like an "aspect", meaning that its a cross-cutting concern over the whole test base. On the other hand, ClosePatient() is an example for a TearDown of maybe one or couple of test fixtures, while other fixtures have their specific TearDown sequences, depending on the test goal.

That's the reason why the base-class approach will not fit our use cases.

One workaround for us could be the IWrapTestMethod interface of the ICommandWrapper. Looks like there the sequential order is fix between the test method and the TearDown. However we saw that it cannot be applied on Fixture or assembly Level (see https://github.com/nunit/nunit/issues/3022). That would be a too huge impact on a test base with >100k test cases. Does the issue have a chance to get processed? Would a contribution from our side be helpful or are there design considerations which are blocking it currently?

**Comment by CharliePoole** [comment]
I was trying to describe what can be done with SetUp and TearDown in the existing framework. For changes in the framework, I'll leave it to the folks who currently work on it. :-)

I do suggest that you come up with something that works for you even if it's not scalable. Then your request to the team would be
in the nature of "Could you make this more scalable, by allowing it to be defined at the assembly level." From what you have described so far, the non-scalable solution doesn't need any more than SetUp and TearDown.

**Comment by TSAVogt** [comment]
Hi,

I am working in a sibling organization unit as @z002Holpp and I am confronted with the same issue.

@CharliePoole: besides the discussion about a solution in the framework I would like to challenge and understand why this cannot be solved by extending the engine.

I found the following flag in the engine: [SynchronousEvents](https://github.com/nunit/nunit-console/blob/d5d6fd6a15e2f082a8b80ab0b21bbae7ac6a90a1/src/NUnitConsole/nunit-console/FrameworkPackageSettings.cs#L78)

When taking the prototype above, extending the engine and setting this flag to true I was also able to execute and process test events synchronously. So, it would exactly solve our problem.

However, as this flag is not documented and easily accessible from outside it does not seem to be a stable or reliable feature of NUnit. It would be a high risk for us if we would use this hidden API and it is removed later on.

Just for understanding: this multi test agent scenario you mentioned [here](https://github.com/nunit/nunit/issues/4193#issuecomment-1275418412) is needed for parallel test execution, right?
As we need synchronous execution only for higher level tests parallel test execution is here anyway out of our scope.

So, what do you think, could this be a reliable way for us to solve our problem?

**Comment by CharliePoole** [comment]
The SynchronousEvents flag was put in place for use by Microsoft's Test Explorer and deliberately left undocumented.

First, let me get out of the way the fact that I opposed the change. However, we operate democratically in the NUnit Project and others felt that the NUnit3 Adapter used by Test Explorer was rapidly becoming the most important client of the engine.

The change we made at the time was as follows:

1. The engine would not create an AppDomain as if `--domain:None` had been specified. This is a feature of the console runner, which we have discouraged folks from using because it requires all our code as well as the test assembly to be present in the same directory. In this case, Microsoft was providing an AppDomain, which was guaranteed to work. I don't know how this will work under version 4.0 of the engine, which no longer supports running tests without a separate agent process. UPDATE: This change is processed in the engine, based on a setting passed by the NUnit3 adapter.

2. Events would be "synchronous." Note that events are not currently only asynchronous from the point of view of the originating code inside the engine. They are managed in a queue in such a way that only one event may be processed at a time. In this case, "Synchronicity" meant that only one event would be processed at a time and no other events could take place until the processing completed. UPDATE: This change is processed by the framework, based on a setting passed by the NUnit3 adapter.

As the person maintaining the engine, I find this a scary feature. What Microsoft does with these events is something of a trade secret and it hasn't been shared with me. @OsirisTerje who maintains the NUnit3 adapter may have received more info, although I imagine it's under NDA.

That's probably enough discussion of the engine for this place. If you would like to go into it further, you can write me offline or create an engine issue. In either case, you should try to come up with a fairly precise definition of what "synchronous" would mean for your purposes.

UPDATE: If all you need is some assurance that the framework will honor the `SynchronousEvents` flag in the future, this is the place to discuss it. I no longer have a role in framework development, so it would be up to @rprouse and others who work on it framework to let you know whether this API will continue to exist in version 4. If you also need the feature to be exposed by `nunit3-console.exe` lets do that as a separate issue.

**Comment by CharliePoole** [comment]
@TSAVogt I see I didn't answer your question about parallel execution.

The engine supports running multiple assemblies in parallel using separate agent processes. The framework supports running individual tests within an assembly in parallel using multiple threads. Both kinds of parallel execution may co-exist so, in the worst case, the engine may have to deal with events coming in any order from various threads in various processes.

Currently, the engine deals with most events at a high level in whatever order they arrive. It would be possible to segregate events from different processes by dealing with them at a lower level but that's a fairly significant change, which would require a significant benefit to make it worthwhile.

I hope that makes it clearer. Let's continue either in the engine project or offline.

**Comment by CharliePoole** [comment]
Note updates to my previous comment!

**Comment by TSAVogt** [comment]
Thanks a lot for the input!

Then we have now to two possible ways to go. The synchronous event discussion we will start in the engine project and in parallel we will also prototype some ICommandWrapper solution. We will come back once we have new insights.
 
@CharliePoole: thanks for the suggestion for an offline discussion. This would make things easier to discuss. Maybe we could go for it when we got some results from the prototype. I will approach you by your gmail address. Fine?

**Comment by CharliePoole** [comment]
Great!
